src/test/java/com/devonfw/sample/archunit/ThirdPartyRulesTest.java

```diff
@@ -1,31 +1,244 @@
+    }
+    return false;
+  }
+
+  private static boolean isImplementingHibernateEnversInternalsDirectly(String targetPackageName) {
+
+    if (targetPackageName.startsWith(ORG_HIBERNATE_ENVERS) && targetPackageName.contains("internal")) {
+      return true;
+    }
+    return false;
+  }
+
+  private static boolean isUsingHibernateOutsideOfImplScope(JavaClass source, String targetPackageName) {
+
+    PackageStructure sourcePkg = PackageStructure.of(source);
+    if (!sourcePkg.isScopeImpl() && targetPackageName.startsWith("org.hibernate")
+        && !targetPackageName.startsWith(ORG_HIBERNATE_VALIDATOR)) {
+      return true;
+    }
+    return false;
+  }
+
+  private static boolean isNotImplementingHibernateEnversInImplScope(JavaClass source, String targetPackageName,
+      String targetPackageFullName, String targetSimpleName) {
 
-        @ArchTest
-        private static ArchRule check_object_dependency = noClasses()
-                        .should().dependOnClassesThat()
-                        .haveFullyQualifiedName("com.google.common.base.Objects")
-                        .because("Use Java standards instead (java.util.Objects).");
+    PackageStructure sourcePkg = PackageStructure.of(source);
+    if (!sourcePkg.isScopeImpl() && targetPackageName.startsWith(ORG_HIBERNATE_ENVERS)
+        && (!targetPackageFullName.equals(ORG_HIBERNATE_ENVERS) || targetSimpleName.startsWith("Default")
+            || targetSimpleName.contains("Listener") || targetSimpleName.contains("Reader"))) {
+      return true;
+    }
+    return false;
+  }
 
-        @ArchTest
-        private static ArchRule check_converter_dependency = noClasses()
-                        .should().dependOnClassesThat()
-                        .haveFullyQualifiedName("javax.persistence.Convert")
-                        .because("Use the javax.persistence.Converter annotation on a custom converter"
-                                        + " which implements the javax.persistence.AttributeConverter instead of the 'javax.persistance.Convert' annotation");
+  public static final ArchCondition<JavaClass> misUseHibernate = new ArchCondition<JavaClass>(
+      "misuse hibernate (Rule-E4).") {
+    @Override
+    public void check(JavaClass source, ConditionEvents events) {
 
-        @ArchTest
-        private static ArchRule check_mysema_dependency = noClasses()
-                        .should().dependOnClassesThat().resideInAPackage("com.mysema.query..")
-                        .because("Use official QueryDSL (com.querydsl.* e.g. from com.querydsl:querydsl-jpa).");
+      for (Dependency dependency : source.getDirectDependenciesFromSelf()) {
+        String targetPackageName = dependency.getTargetClass().getPackageName();
+        String targetPackageFullName = dependency.getTargetClass().getFullName();
+        String targetClassDescription = dependency.getDescription();
+        String targetSimpleName = dependency.getTargetClass().getSimpleName();
```

Comment:
FYI: why not just passing the `Dependency` or `JavaClass` as argument to the individual private check methods?
Passing the entire structured information is more semantic and stable that passing only `String` attributes.

Reply:
I adjusted the private check method signatures to use JavaClass parameters.